### DISCUSSION TITLE
AsyncIO Code works in Python code but not in MicroPython

### DISCUSSION THREAD

**Discussion by neliel123** [micropython/micropython#12950] [General]
I have just started exploring MicroPython specifically AsyncIO concepts.

I have written this code and tried running it in both CPython and MicroPython environments using the latest firmware but I don't know why the output is different.
```
import asyncio

sample_global = 0

async_event = asyncio.Event()

async def update_globals():
    global sample_global
    print('Function Started...')

    await async_event.wait()
    print('Sleep...Before Updating...')

    await asyncio.sleep(1)
    print('Updating...')

    sample_global += 1
    print('Done Updating...')

async def main():

    tasks = [update_globals() for _ in range(5)]

    asyncio.gather(*tasks)

    await asyncio.sleep(2)  # Simulating some time before allowing updates
    async_event.set()

    await asyncio.sleep(2)
    print(f"Final value of sample_global: {sample_global}")

if __name__ == "__main__":
    asyncio.run(main())
```

The output in CPython is `Final value of sample_global: 5`
But in MicroPython it is `Final value of sample_global: 0`

Any hints?

**Comment by peterhinch** [comment]
`gather` is a coroutine. The following works under CPython and MP:
```python
import asyncio

sample_global = 0

async_event = asyncio.Event()

async def update_globals():
    global sample_global
    print('Function Started...')

    await async_event.wait()
    print('Sleep...Before Updating...')

    await asyncio.sleep(1)
    print('Updating...')

    sample_global += 1
    print('Done Updating...')

async def foo(tasks):
    await asyncio.gather(*tasks)

async def main():

    tasks = [update_globals() for _ in range(5)]
    asyncio.create_task(foo(tasks))

    await asyncio.sleep(2)  # Simulating some time before allowing updates
    async_event.set()

    await asyncio.sleep(2)
    print(f"Final value of sample_global: {sample_global}")

if __name__ == "__main__":
    asyncio.run(main())
```

**Reply by neliel123** [reply]
>gather is a coroutine. 

Hi Peter, 
Maybe I don't get it that much.  Many of the tutorials I am following are in CPython format as there are more tutorials about AsyncIO in there.

Is this a MicroPython quirk so it does not work as per the CPython tutorials that I have been following on the net?  This is so that I could take note of this.
Thanks for replying

**Reply by peterhinch** [reply]
See [CPython docs](https://docs.python.org/3/library/asyncio-task.html#running-tasks-concurrently) where `gather` is defined as an `awaitable`. The original code has a bug where `gather` is called as if it were a function. With this bug fixed, MP and CPython behave identically. At the moment I can't see a `quirk`.

For a MicroPython asyncio tutorial see [this doc](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md).

**Reply by neliel123** [reply]
Hey,
So if gather is an awaitable then I could await on it?
```
async def main():

    tasks = [update_globals() for _ in range(5)]

    await asyncio.gather(*tasks)

    await asyncio.sleep(2)  # Simulating some time before allowing updates
    async_event.set()

    await asyncio.sleep(2)
    print(f"Final value of sample_global: {sample_global}")
```
Tried this, and I am seeing a different output.
```
Function Started...
Function Started...
Function Started...
Function Started...
Function Started...
```
It did not print the final value.

**Reply by peterhinch** [reply]
The tasks have started but are waiting on the `Event`. The `gather` only terminates when all the tasks have terminated, so the `Event` never gets triggered. The code therefore waits forever. That's why I delegated `await gather(...)` to a concurrent task (`foo()`).
